extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -830,4 +884,559 @@ private Table readTableInternal(
+        } else {
+            throw new IllegalArgumentException(
+                    "Table does not exist: " + tableIdentifier + ", update the instructions " +
+                            "to create the table if it does not exist and try again.");
+        }
+
+        try {
+            final List<CompletedParquetWrite> parquetFileinfo =
+                    writeParquet(icebergTable, dhTables, writeInstructions);
+            final List<DataFile> appendFiles = dataFilesFromParquet(parquetFileinfo);
+            if (addSnapshot) {
+                commit(icebergTable, newSpecAndSchema, appendFiles, writeInstructions, overwrite && !newTableCreated);
+            }
+            return appendFiles;
+        } catch (final RuntimeException writeException) {
+            if (newTableCreated) {
+                // Delete it to avoid leaving a partial table in the catalog
+                try {
+                    catalog.dropTable(tableIdentifier, true);
+                } catch (final RuntimeException dropException) {
+                    writeException.addSuppressed(dropException);
+                }
+            }
+            if (newNamespaceCreated) {
+                // Delete it to avoid leaving a partial namespace in the catalog
+                dropNamespaceIfExists(tableIdentifier.namespace());
+            }
+            throw writeException;
+        }
+    }
+
+    /**
+     * Check if the schema for the iceberg table is compatible for overwriting with a deephaven table with provided
+     * definition.
+     */
+    private static void verifyOverwriteCompatibility(
+            final Schema icebergSchema,
+            final Schema newSchema) {
+        if (!icebergSchema.sameSchema(newSchema)) {
+            throw new IllegalArgumentException("Schema mismatch, iceberg table schema: " + icebergSchema +
+                    ", schema derived from the table definition: " + newSchema);
+        }
+    }
+
+    /**
+     * Check if the partition spec for the iceberg table is compatible for overwriting with a deephaven table with
+     * provided definition.
+     */
+    private static void verifyOverwriteCompatibility(
+            final PartitionSpec icebergPartitionSpec,
+            final PartitionSpec newPartitionSpec) {
+        if (!icebergPartitionSpec.compatibleWith(newPartitionSpec)) {
+            throw new IllegalArgumentException("Partition spec mismatch, iceberg table partition spec: " +
+                    icebergPartitionSpec + ", partition spec derived from table definition: " + newPartitionSpec);
+        }
+    }
+
+    /**
+     * Check if the schema for the iceberg table is compatible for appending a deephaven table with provided definition.
+     */
+    private static void verifyAppendCompatibility(
+            final Schema icebergSchema,
+            final TableDefinition tableDefinition) {
+        // Check that all columns in the table definition are part of the Iceberg schema
```

Comment:
Is it an error to write a table with extra columns not in Iceberg schema? Or should we only write the matching columns?

Reply:
I have added a iceberg instruction for verify compatability, if the user wants to verify that the data being appended/overwritten is compatible with the original table.

- For appending, we check that all required columns are present in the data with compatible types, and no extra columns outside of the schema are present.
-For overwriting, we check if the schema is identical.

If the user wants to override these checks, they can disable them through iceberg instructions.
I can add more details in the comments for the new iceberg instruction.

Reply:
It's the `and no extra columns outside of the schema are present` test that concerns me. We will be requiring a user to `dropColumns()` to meet this compatibility metric when I'm not sure it's important at all.

`compatible` is pretty broad in definition (IMO), should not mean `identical`.

Reply:
From the perspective of DH controlling the writing, I think we can be opinionated, and prefer to give the user less control than we might otherwise want / need to. **I don't think it makes sense to allow the user to specify they want to write out Deephaven columns to the parquet file that _aren't_ mapped to the Iceberg table**. By default, it may be appropriate to always use the latest Schema at the time it is being written to, but I think we need to **allow the user to pick the Schema they want to use for writing**. IMO, the physical parquet columns we write should be a (non-strict) subset of that Schema's columns. If there is a map between a DH column and a Schema column, we write it to parquet; otherwise, we exclude it. This also means that _every_ column we write out in this way has an Iceberg field_id we can map into parquet's field_id.

Reply:
`IMO, the physical parquet columns we write should be a (non-strict) subset of that Schema's columns.`
The extra check that I have right now is that all the required columns should be there in the tables being appended.